### DISCUSSION TITLE
Real time clock

### DISCUSSION THREAD

**Discussion by Pucksnok** [missionpinball/missionpinball#4] [Ideas]
Being able to use computer's clock as a real time clock where we would be able to use the current date/time in MPF would be great.
I could see myself using this for a automatic night mode with lower sounds, a midnight mode, date and time (timestamp) for high scores etc.

**Comment by colemanomartin** [comment]
I don't see any documentation that MPF has access to the system clock but Python does.  I think you would need to write some mode code that would grab the system time and assign a machine variable to the value.  this might get you started on the Python side.  https://www.programiz.com/python-programming/datetime/current-datetime
Jan did a youtube video on writing mode code.  If you get it working, please post back.  It would also make a nice addition to the cookbook part of the docs.

**Reply by Pucksnok** [reply]
Thx, I may try and learn some python later on when work and familylife is less hectic. 🥳

**Reply by Pucksnok** [reply]
I'm trying to learn, no prior programming skills makes it hard. I have now done a mode with python code so it writes a machine variable with the current time. So far it works. My issue now is when I do a infinity loop to write the time every second everything else stops working, my guess is that it's because of how python works in some way, it won't continue with everything else when in the loop, no multitasking here apparently. ;) I do see the event when changing the time variable every second so that apparently works at least.

I guess I could move around it with setting up a timer in yaml that post a event every second, or every 30th second (hh:mm should be enough). Now Ijust need to figure out how I use "def" when a event happen instead of start of mode.

**Reply by Pucksnok** [reply]
I now have it working. Instead of using an infinite loop in my python code I use a timer to trigger an event every 10 seconds that makes the python code to save the time to a machine variable. I use restart_on_complete for the timer to make it write every 10th second. Works great.

Thanks @colemanomartin for guiding me where to start.

**Reply by colemanomartin** [reply]
That's fantastic!  Can you post the yaml for the timer and the mode code you used?

**Reply by Pucksnok** [reply]
My intention were to publish my whole machine when done (we know it's never completely done...) but I published it on my github now instead.

You find it here, I'm no coder so if you spot something strange, please let me know. :)
https://github.com/Pucksnok/clock_mpf

**Comment by Pucksnok** [comment]
Published my solution here >> https://github.com/Pucksnok/clock_mpf << if anyone need it, I'm no coder so please comment if you see something strange.
